File: mobile_app/lib/models/application_tracker.dart
class ApplicationTrackerItem {
  ApplicationTrackerItem({
    required this.id,
    required this.smeId,
    required this.productName,
    required this.productType,
    required this.status,
    this.notes,
    required this.createdAt,
    required this.updatedAt,
  });

  final int id;
  final int smeId;
  final String productName;
  final String productType;
  final String status;
  final String? notes;
  final DateTime createdAt;
  final DateTime updatedAt;

  factory ApplicationTrackerItem.fromJson(Map<String, dynamic> j) => ApplicationTrackerItem(
        id: j['id'] as int,
        smeId: j['sme_id'] as int,
        productName: j['product_name'] as String,
        productType: j['product_type'] as String,
        status: j['status'] as String,
        notes: j['notes'] as String?,
        createdAt: DateTime.parse(j['created_at'] as String),
        updatedAt: DateTime.parse(j['updated_at'] as String),
      );
}

class ApplicationTrackerResponse {
  ApplicationTrackerResponse({required this.smeId, required this.applications});

  final int smeId;
  final List<ApplicationTrackerItem> applications;

  factory ApplicationTrackerResponse.fromJson(Map<String, dynamic> j) =>
      ApplicationTrackerResponse(
        smeId: j['sme_id'] as int,
        applications: (j['applications'] as List<dynamic>? ?? [])
            .map((e) => ApplicationTrackerItem.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

File: backend/app/models/application_tracker.py
"""Application Tracker — persists SME financing application status."""

from __future__ import annotations

from datetime import datetime, timezone

from sqlalchemy import DateTime, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


def _now() -> datetime:
    return datetime.now(timezone.utc).replace(tzinfo=None)


class ApplicationTracker(Base):
    __tablename__ = "application_tracker"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    sme_id: Mapped[int] = mapped_column(Integer, nullable=False, index=True)
    product_name: Mapped[str] = mapped_column(String(255), nullable=False)
    product_type: Mapped[str] = mapped_column(String(64), nullable=False)   # bnpl | micro_credit | grant
    status: Mapped[str] = mapped_column(String(64), default="draft")        # draft|submitted|under_review|approved|rejected
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=_now)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=_now, onupdate=_now)
